OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnMovement.cs

```diff
@@ -0,0 +1,68 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class UpgradeOnMovementInfo : UpgradableTraitInfo, Requires<UpgradeManagerInfo>, Requires<IMoveInfo>
+	{
+		[UpgradeGrantedReference, FieldLoader.Require]
+		[Desc("The upgrades to grant.")]
+		public readonly string[] Upgrades = { };
+
+		public override object Create(ActorInitializer init) { return new UpgradeOnMovement(init.Self, this); }
+	}
+
+	public class UpgradeOnMovement : UpgradableTrait<UpgradeOnMovementInfo>, ITick
+	{
+		readonly UpgradeManager manager;
+		readonly IMove movement;
+
+		bool granted;
+		bool isMoving;
+		WPos cachedPosition;
+
+		public UpgradeOnMovement(Actor self, UpgradeOnMovementInfo info)
+			: base(info)
+		{
+			manager = self.Trait<UpgradeManager>();
+			movement = self.Trait<IMove>();
+			cachedPosition = self.CenterPosition;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (IsTraitDisabled)
+				return;
+
+			var oldCachedPosition = cachedPosition;
+			cachedPosition = self.CenterPosition;
+
+			isMoving = !self.IsDead && (movement.IsMoving || (oldCachedPosition - cachedPosition).HorizontalLengthSquared != 0);
```

Comment:
During my tests there were some 'hiccups' when only using IMove.IsMoving (during turns, usually). If we managed to make that more robust instead this could go away, but until then this is needed to make sure the upgrades are applied all the time while moving.